### PR TITLE
Build hdf5 using CMake instead of Autogen

### PR DESCRIPTION
**Pull request by TotoGaz**
This PR avoids the spurious detection of the `NDEBUG` compilation flag by `CMake` when importing the `hdf5` package.

As a side-effect, it will allow to test the `GEOS` asserts in github-ci configs running in `Debug` mode.

See:
  - `GEOS` issue https://github.com/GEOS-DEV/GEOS/issues/2631
  - `hdf5` issue https://github.com/HDFGroup/hdf5/issues/3526#issuecomment-1719664682

I initially proposed this PR from a fork of the `TPL`s but the ci fail due to missing authorization to connect to DockerHub.
Previous discussion is available in the [initial PR](https://github.com/GEOS-DEV/thirdPartyLibs/pull/239).

Reverts GEOS-DEV/thirdPartyLibs#247 which was reverting https://github.com/GEOS-DEV/thirdPartyLibs/pull/243

closes https://github.com/GEOS-DEV/GEOS/issues/2631
Relates to https://github.com/GEOS-DEV/GEOS/pull/2721/

### CODE REVIEW

**Comment by TotoGaz** [comment]
@Algiane @tbeltzun I'm sorry I had to revert your PR https://github.com/GEOS-DEV/thirdPartyLibs/pull/243: our policy is that `geos` always points to the HEAD of the TPL, which was not the case since https://github.com/GEOS-DEV/GEOS/pull/2721/ is not merged. PR https://github.com/GEOS-DEV/thirdPartyLibs/pull/243 should not have been merged.

As a result, anyone cloning both `geos` and the TPLs would not be able to build. Furthermore, it was preventing any other upgrade of the TPL.

I've reopened the current PR (#248) and updated https://github.com/GEOS-DEV/GEOS/pull/2721 with the new TPL tag so everything should be ready on your side (no additional work I hope). Thank you for your understanding!

**Comment by Algiane** [comment]
In last commits, we:
  - switch to `RelWithDebInfo` compilation for `hdf5` if user asks for `Debug` mode: it is a workaround to allow `Silo`'s `Autogen` script to find the `hdf5`library (see [this issue](https://github.com/LLNL/Silo/issues/369) ). Note that before this PR, `hdf5` was always build in `Release` mode. 
Once the implementation of `Silo's` build with `CMake` will be ready and released, it will be possible to link with the `libhdf5_debug` library;

   - solve compilation errors due to `implicit declaration of function 'zfp_init_zfp'` (and missing semi-colons) when building `Silo'`s release 4.11 with `Clang` (see [closed issue](https://github.com/LLNL/Silo/issues/224)).  For now, I have added a `-w` flag to compiler flags if `Clang` is detected (prior this PR, when building `hdf5` using `Autotools`, this flag was automatically forwarded inside the `HDF5_C_FLAGS` variable). It allows to build `Silo` on all the architectures covered by the current ci and on OSX.
Such a fix is not very satisfying (it probably doesn't work with Clang on windows and we may have issues with other compilers or on exotics architectures): as the compilation errors are solved in Silo's release `4.11.1`. I will propose an upgrade to this latter and the removal of the conditional adding of ` -w` flag in another PR (once the current one will be merged).

File: CMakeLists.txt [hunk 1]
```diff
@@ -278,6 +294,19 @@ set(SILO_DIR "${CMAKE_INSTALL_PREFIX}/silo")
...
+# AppleClang or regular Clang due to implicit declaration of
+# 'zfp_init_zfp'. This issue is fixed by release 4.11.1 so it will be
+# possible to remove the `-w` flag when we will use this latter (see
+# https://github.com/LLNL/Silo/issues/224).
+#
+# This -w flag was exported inside the HDF5_C_FLAGS variable when hdf5 was built
+# using autogen.
+#
+if( CMAKE_CXX_COMPILER_ID MATCHES "Clang" )
+  # HDF5_C_FLAGS is used only here so we can overwritting it
```
Comment (TotoGaz) [inline CMakeLists.txt:307]:
```suggestion
  # HDF5_C_FLAGS is used only here so we can overwrite it
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -1258,7 +1287,19 @@ if( ENABLE_FESAPI )
...
-                                    -DWITH_TEST:BOOL=OFF)
+                                    -DWITH_TEST:BOOL=OFF
+                                    # Next lines prevent compilation
+                                    # errors due to an issue with the
+                                    # detection of the hdf5 version in
+                                    # fesapi < 2.4.0.0 (on CentOS, for
+                                    # Release build of hdf5, see
+                                    # https://github.com/F2I-Consulting/fesapi/issues/326)
+                                    # It should be possible to remove it when
+                                    # fesapi-2.9.0.0 will be released
```
Comment (TotoGaz) [inline CMakeLists.txt:1298]:
```suggestion
                                    # Next lines prevent compilation errors due to an issue
                                    # with the detection of the hdf5 version in fesapi < 2.4.0.0
                                    # (on CentOS, for Release build of hdf5, see https://github.com/F2I-Consulting/fesapi/issues/326)
                                    # It should be possible to remove it when fesapi-2.9.0.0 will be released
```

**Review by TotoGaz — COMMENTED** [review]
LGTM.
@Algiane you can approve it yourself! I cannot do it since I've reopened  it, it's mine and github does not want me to approve my PR!

**Comment by Algiane** [comment]
Hi @TotoGaz ,

Thanks, I will approve the current PR asap!

From the TPL side I think that everything is OK but there are still some issues when building Geos in Debug with Cuda on Ubuntu (we cannot remove the "ready to be merged" flag as Cuda jobs are not launched without this flag).

Best

**Review by Algiane — APPROVED** [review]

**Comment by Algiane** [comment]
Hi @TotoGaz ,

I think that both PR are ready to be merged now : 
  - the ci is fixed for [GEOS PR 2721](https://github.com/GEOS-DEV/GEOS/pull/2721) that builds GEOS against the image constructed from the current branch;
  - the current branch builds the TPLs without errors for the ci and on OSX.
  
 Thank you by advance for your help to merge this work.
 Best,

**Pull request merged**